### PR TITLE
massdrop alt/ctrl: support saving into nvm

### PR DESCRIPTION
**Pull request by daltona**
Use SAMD51 virtual eeprom to store eeprom in nvm instead of ram buffer so it is persistent accross reboots.
In order to have this working it is required to set PSZ and SBLK values in the NVM user row, I believe that can be done with the mdloader, but not having the source code, I am not able to implement this, I would be happy to do so.
I've tested by manualy updating the NVM user row connecting a JLINK probe on my keyboard.

Signed-off-by: Alexandre d Alton <alex@alexdalton.org>

### CODE REVIEW

**Comment by drashna** [comment]
@patrickmt you should absolutely see this.

**Comment by drashna** [comment]
Anyone want to test this out? 

@XScorpion2 @reywood  @abishalom @MatthewRobo @valen214

**Comment by reywood** [comment]
I wouldn’t be able to test the nvm part, but could test that it behaves normally without those bits set.

**Comment by XScorpion2** [comment]
I'll test this out later tonight I hope

**Comment by zvecr** [comment]
To fully understand the situation, does the proposed change require mdloader changes or external hardware to function? If this code was merged as-is, what would users who dont have access to the above see? If its runtime failures, I dont see how this can be merged yet.

**Comment by reywood** [comment]
@zvecr Looks like the code will work with or without changes to mdloader. Though you wouldn't see the benefits until mdloader is changed.

**Comment by zvecr** [comment]
@reywood that's good. Let's see what the testing brings.

**Comment by XScorpion2** [comment]
Well, so far my CTRL has yet to crap itself from these changes. Also types just fine, except my fingers are now too use to a grid layout...

**Comment by drashna** [comment]
@patrickmt

**Comment by daltona** [comment]
I can try to add the code that will automatically update the configuration area, but this is too dangerous in my opinion as it cannot be recovered without a change in mdloader or JTAG hardware.

having mdloader changed would allow to securely update the configuration area without risking to lose some important data that is stored in it.

**Comment by XScorpion2** [comment]
Please no, last thing that needs to be floating around is test code that could break a popular kb. We can wait for an official mdloader change.

**Comment by daltona** [comment]
fully agree.
Well I figured out that there is host part of mdloader in the form of source code, so I might be able to implement the NVM configuration change in here.
I'll implement this, try and report back.

**Comment by daltona** [comment]
This patch to mdloader allows to enable smarteeprom feature in order to test configuration persistence.
https://github.com/daltona/mdloader/commit/0aaf775edede8e34dc511037e3a8cc27e07c7782
tested on a mac.

**Comment by XScorpion2** [comment]
@daltona tried your modified mdloader with no luck. It still does not save out rgb matrix state. Did I use the right option?
`.\mdloader_windows.exe -f -D -r .\massdrop_ctrl_xulkal.bin --restart`

**Comment by drashna** [comment]
I believe that this closes #4625

**Comment by XScorpion2** [comment]
Any word on getting this into master?

**Comment by drashna** [comment]
Well, there is a merge conflict present, so that needs to be resolved first.

**Comment by daltona** [comment]
> Well, there is a merge conflict present, so that needs to be resolved first.

I've rebased and fixed the conflict.

**Comment by drashna** [comment]
Just wanted to make sure, this has no issues if the change from mdloader isn't made, it just retains the existing behavior.  Correct?

**Comment by daltona** [comment]
> Just wanted to make sure, this has no issues if the change from mdloader isn't made, it just retains the existing behavior. Correct?

This is correct, if the virtual eeprom is not enabled by the mdloader change, the code behave the exact same as before.

**Comment by drashna** [comment]
@daltona could you open a PR to massdrop's or patrickmt's repo for mdloader?  
If the changes to mdloader don't occur, it may not be worth having this code change merged in

**Comment by XScorpion2** [comment]
poke, bad bot =p

**Comment by zvecr** [comment]
Not really bad bot, as no one has added support to the flasher. If it's not going to happen, personally I wouldn't mind if this gets closed.

**Comment by daniel5151** [comment]
If it helps at all, I've been using this PR (along with the updated flasher) for the past couple months, and it's been working swimmingly.

It'd be a real shame not getting this merged in.

**Comment by XScorpion2** [comment]
I've been using this pr and updated flasher as well just fine. Having the code merged in, but unused due to the flasher not being updated isn't entirely a bad thing.

**Comment by daltona** [comment]
I agree, It can't harm to have this code in.
In the mean time, I am cleaning up the PR for the loader so it should progress soon.

**Comment by reywood** [comment]
I'll add my wish that this be merged to the pile.

**Comment by zvecr** [comment]
The issue is a maintenance one, having "dead" code, that you can only utilise if you compile a tool from a random fork. It can make something like the work within #7274 more complicated. Having to cherry pick stuff to get the flasher to work, isnt really that different to having to cherry pick this PR.

**Comment by drashna** [comment]
@just-another-jxliu @pschmuckal

**Comment by pschmuckal** [comment]
@daltona Thank you for doing this.

I finally had a chance to review.  My comments:
- Needs to be rebased again as eeprom.c has been corrected wrt EEPROM_SIZE which now must be EECONFIG_SIZE (34by) or larger.
- Please use mfgr reg defines in samd51j18a.h.  i.e. line 23 should be
    ```volatile uint8_t *SmartEEPROM8 = (uint8_t *) SEEPROM_ADDR;```
- Be great if we could do address bounds checking so we don't accidentally write to non-existent memory.
- If a busy timeout occurs, we should abort, not ignore and attempt to read/write to NVM anyway.
- Actual SmartEEPROM size is much larger (currently 2048 by) than RAM array.  Should we allow writing to the larger EEPROM memory addrs space?  Limit by PSZ.

Now to the bigger issue:  The standard method in this case would be to detect if the EEPROM has not yet been configured during an startup/init routine, and then just automatically configure it to default values.  This makes it automatically work and no modifications to mdloader are required.

However, I understand the hesitancy to do this, and the work (& risk) to modify the NVM User Page automatically is considerable - particularly if the action takes place during powerup where power may not be stable.  And there is other mfgr config/calibration settings close by that could get damaged.

It also complicates manufacturing somewhat as another step is required to configure this.

I would probably still lean towards the automatic initialization, but I'm ok with the mdloader approach as well.

**Comment by daltona** [comment]
@pschmuckal
> * Be great if we could do address bounds checking so we don't accidentally write to non-existent memory.

What should we do if invalid ? What to return in case of a read out of bounds ? will 0xFF be ok ?

> * If a busy timeout occurs, we should abort, not ignore and attempt to read/write to NVM anyway.

Same as above

> * Actual SmartEEPROM size is much larger (currently 2048 by) than RAM array.  Should we allow writing to the larger EEPROM memory addrs space?  Limit by PSZ.

OK

> It also complicates manufacturing somewhat as another step is required to configure this.

I think it can be easily done at manufacturing if it is done in a controlled way right after flashing. It is quite trivial to create an OpenOCD / JLINK Flash script that will update the NVM User Page at a safe time.

**Comment by pschmuckal** [comment]
@daltona My preference would be to return 0x00 in case of error as that is less like to cause strange behavior.

**Comment by pschmuckal** [comment]
@daltona Looks good. Thanks.

**Comment by drashna** [comment]
@pschmuckal if this looks good, would you mind approving it?  Having somebody that is more familiar with the ATSAM code would go a long way.

**Comment by tzarc** [comment]
Is there any particular reason the SEESBLK/SEEPSZ can't be written to by QMK's startup code?
A quick scan of the datasheet shows that NVMCTRL could be used to do so, but the values aren't available until a reset occurs.  Would it be worth attempting to init during startup, and then the reliance on mdloader isn't required?

**Comment by tzarc** [comment]
One other point of note: invariably someone is going to want to enable VIA support for these boards -- as long as we have the ability to erase/resize based on whatever the maximum size required is (i.e. overriding EEPROM_SIZE), then I figure we should be okay.

**Comment by drashna** [comment]
> Is there any particular reason the SEESBLK/SEEPSZ can't be written to by QMK's startup code?
> A quick scan of the datasheet shows that NVMCTRL could be used to do so, but the values aren't available until a reset occurs. Would it be worth attempting to init during startup, and then the reliance on mdloader isn't required? The last useful commit to mdloader was in 2018.

My biggest concern here would be that it may be easy to create a situation where a bootloop was esy to trigger.

**Comment by tzarc** [comment]
Parking this at the moment until we get a merge on Massdrop/mdloader#16.

**Comment by nguyer** [comment]
I just wanted to say, this is awesome! I've been wanting this feature since I bought my CTRL board when they were first released.

One question... is it possible to also persist the current active layer as well? I have a layer that toggles back and forth between "mac" and "default" layouts, and it would be nice to not have to set that every time it powers up.

Thanks again!

**Comment by kaeltis** [comment]
Working perfectly for me as well, I really hope will make it into the master branch soon :)
Since the official mdloader repo and it's maintainers seem to have been inactive for quite some time, maybe it would be best to maintain an "official" community fork of it?

**Comment by skullydazed** [comment]
> Since the official mdloader repo and it's maintainers seem to have been inactive for quite some time, maybe it would be best to maintain an "official" community fork with prebuilt binaries for now?

If someone took on that work it would be a great resource for the community of people with drop boards.

We've talked a little about doing that as part of QMK but right now there doesn't seem to be anyone who wants to take on the work. If someone created and started to maintain a community package we'd be very happy to use and support their ongoing work, but at this time it's not a project we have the bandwidth to handle.

**Comment by noahgaertner** [comment]
> > Since the official mdloader repo and it's maintainers seem to have been inactive for quite some time, maybe it would be best to maintain an "official" community fork with prebuilt binaries for now?
> 
> If someone took on that work it would be a great resource for the community of people with drop boards.
> 
> We've talked a little about doing that as part of QMK but right now there doesn't seem to be anyone who wants to take on the work. If someone created and started to maintain a community package we'd be very happy to use and support their ongoing work, but at this time it's not a project we have the bandwidth to handle.

Hey, I can look into maintaining a fork, but I'm a fairly new C dev and I'd need someone else to double check any pull requests just to make sure I don't miss any bad code or stuff that will break something.

**Comment by YoungKey83** [comment]
> Working perfectly for me as well on a CTRL, I really hope this will make it into the master branch soon :)
> 
> Since the official mdloader repo and it's maintainers seem to have been inactive for quite some time, maybe it would be best to maintain an "official" community fork with prebuilt binaries for now?
> It could be referenced in the QMK documentation and keyboard readmes.

Hello, I also use CTRL now and want to save LED. First time I just thought eeprom.c is only thing that I have to change but it's not working. Could you please let me know how to do that? Thanks.

**Comment by tomatsaev** [comment]
Any updates on this? Quite lost with all the different branches, is there a guide maybe on enabling it? Thanks.

**Comment by drashna** [comment]
> Any updates on this? Quite lost with all the different branches, is there a guide maybe on enabling it? Thanks.

The problem is, this change doesn't mean anything without the official mdloader flasher supporting it.  And if we merge this in without that support, then the support for it will fall to us to provide, basically. 

So, really, until Massdrop supports it, or somebody decides to pick up support for it, this is basically stuck in limbo. unfortunately.
And given some of massdrop's own PRs ...

**Comment by a7hybnj2** [comment]
I just wanted to make note that I used @daltona s mdloader and qmk on my alt and everything is working perfectly. Thanks!

**Comment by kaeltis** [comment]
Just updated my CTRL to master using this pull request as a patch - still working perfectly. 
(If anyone stumbles across this, download https://github.com/qmk/qmk_firmware/pull/6068.patch and patch using `git am 6068.patch`)

Since https://github.com/Massdrop/mdloader/pull/16 is probably not getting merged any time soon, couldn't the EEPROM enabling be checked/done on QMK startup like @tzarc mentioned? Would there be a high chance of breaking things if massdrop changes anything?

Maybe add it as a flag like `MASSDROP_EEPROM_ENABLE` that's disabled by default to make it opt-in with a disclaimer.

**Comment by a7hybnj2** [comment]
Thought I would leave this here to update my status from yesterday: https://github.com/Massdrop/mdloader/pull/16#issuecomment-723486011

**Comment by TheZoc** [comment]
In hopes to increase the chances of getting this merged, I rewrote Massdrop/mdloader#16 and created a new pull request https://github.com/Massdrop/mdloader/pull/49.

If someone could go in there and review my code, I'd really appreciate.

I also can maintain a fork of mdloader if that's what is required to merge this PR. I'd love to see this get merged :)

**Comment by ewersp** [comment]
Can't wait to have this and the MDLoader PR merged into mainline! Just tested both and can confirm RGB settings are being persisted on my Drop ALT (with a custom keymap) successfully.

Is there any way/hope of moving this forward? It would save everyone hours of reading/searching/debugging to understand what really boils down to a few small changes to enable something that feels like it should be default behavior.

**Comment by fauxpark** [comment]
The ball is really in Massdrop's court with this one, and they don't seem too interested in doing much about it, unfortunately.

**Comment by ash0x0** [comment]
> 
> 
> I can maintain an updated branch of MDLoader with the NVMe support enabled and windows support. Do you think that will be worth it?
> 
> It would be based off [Massdrop/mdloader#49](https://github.com/Massdrop/mdloader/pull/49) and [Massdrop/mdloader#50](https://github.com/Massdrop/mdloader/pull/50)

Given there are "proprietary" blobs involved, I doubt it would be a good idea to maintain a "community official" branch or something similar. Massdrop also seem to have their own unpublished versions of the blob for new keyboards that they're using in their configurator, all of this is in violation of GPL btw. The assumption that you can plug the blobs in and it'd work is false because they change between the MD boards and their revisions so you can never actually maintain that. 
Maintaining a branch under your username (I doubt qmk would want it under the org) would cause confusion too. Even if updated in the keyboard README it'd still be confusing. 
It'd definitely be better than whatever the hell is going on at [Massdrop/mdloader](https://github.com/Massdrop/mdloader) but it's not without issues.
Also, it wouldn't do anything for #11119 which would still hold up this PR.

**Comment by kaeltis** [comment]
How feasible would it be to create our own open source bootloader to be able to ditch their proprietary `applet-mdflash.bin` blob and use the drop keyboards as a "clean slate" for QMK?

**Comment by fauxpark** [comment]
It's certainly possible to write your own bootloader for ATSAM devices, the problem is that you will need extra hardware to replace it, in this case a J-Link and a Tag-Connect cable: https://learn.adafruit.com/how-to-program-samd-bootloaders
image
Which for most users then becomes completely useless once you've done it.

**Comment by kaeltis** [comment]
Ah, I thought the applet (`applet-mdflash.bin`) was the same as the bootloader, thanks for explaining!
This low-level programming still very much seems like witchcraft to me :D

Assuming they are using the official bootloader or something close to it (Bootloader version: v2.18Jun 22 2018 17:28:08, so probably 2.18 from https://www.microchip.com/DevelopmentTools/ProductDetails/PartNO/SAM-BA%20In-system%20Programmer), couldn't we replace the applet and have something very close to a "clean slate" without needing a J-Link?

**Comment by kaeltis** [comment]
I see, thanks for explaining. I thought the `applet-mdflash.bin` was the bootloader, but that makes more sense :)

Apart from the bootloader (that's definitely different between boards, mine is `v2.20 Mar 27 2019 10:04:47 [ctrl]`), would it be of any use if we replaced the `applet-mdflash.bin` with our own? Looking through the mdloader releases, that file hasn't changed at all.

**Comment by fauxpark** [comment]
As far as I can tell, the applet simply contains routines for programming external memories; it is uploaded to the chip before flashing proper starts, so that the bootloader knows where to shove the data it's receiving. From the SAM-BA docs:
> Applets are binary programs that can be loaded through the sam-ba connection into some RAM region of the target before being executed. Applets extend the SAM-BA monitor to provide more features. The main usage of applets is to program the bootstrap into some Non-Volatile Memory (NVM).

So it would make sense that the mdloader's applet file hasn't changed, as it is fairly specific to the hardware configuration. In any case, I don't think it would be useful.

**Review by tzarc — APPROVED** [review]

**Comment by daltona** [comment]
@zvecr Thank you for completing this and allowing it to be merged !

**Comment by zvecr** [comment]
Thanks go to you for putting in the bulk of the work!

**Comment by vmalloc** [comment]
@zvecr @daltona Just to make sure I'm up to speed - what's the bottom line for mdloader? This PR being merged means the qmk side supports persistence, but this can't actually be used until Drop updates their mdloader binaries?

**Comment by nicoelayda** [comment]
@vmalloc mdloader officially supports persistence starting from v1.0.6 https://github.com/Massdrop/mdloader/releases/tag/1.0.6

**Comment by vmalloc** [comment]
@nicoelayda awesome! I missed that bit - updated mdloader and it works great. Thanks for all the hard work you guys put into this!

**Pull request merged**
